source/common/network/win32_socket_handle_impl.cc

```diff
@@ -71,10 +79,43 @@ Api::IoCallUint64Result Win32SocketHandleImpl::recvmmsg(RawSliceArrays& slices,
 }
 
 Api::IoCallUint64Result Win32SocketHandleImpl::recv(void* buffer, size_t length, int flags) {
+  if (flags & MSG_PEEK) {
+    return peek(buffer, length);
+  }
 
-  Api::IoCallUint64Result result = IoSocketHandleImpl::recv(buffer, length, flags);
-  reEnableEventBasedOnIOResult(result, Event::FileReadyType::Read);
-  return result;
+  if (peek_buffer_.length() == 0) {
+    Api::IoCallUint64Result result = IoSocketHandleImpl::recv(buffer, length, flags);
+    reEnableEventBasedOnIOResult(result, Event::FileReadyType::Read);
+    return result;
+  } else {
+    return readFromPeekBuffer(buffer, length);
+  }
+}
+
+Api::IoCallUint64Result Win32SocketHandleImpl::peek(void* buffer, size_t length) {
+
+  // if the `peek_buffer_` has already enough data we can skip reading
+  // from the wire.
+  if (length > peek_buffer_.length()) {
+    // The caller is responsible for calling with the larger size
+    // in cases it needs to do so it can't rely on transparent event activation.
+    // So no in this case we should activate read again unless the read blocked.
+    Api::IoCallUint64Result peek_result = drainToPeekBuffer(length);
+
+    //  Some error happened
+    if (!peek_result.ok()) {
+      if (peek_result.wouldBlock() && file_event_) {
+        file_event_->registerEventIfEmulatedEdge(Event::FileReadyType::Read);
+        if (peek_buffer_.length() == 0) {
+          return peek_result;
```

Comment:
What is the kernel behavior in the case of short peek?  Does it return partial data?
actually would it make sense to move the peek tests to common socket handle file so we can test the behavior matches?

Reply:
Yes, I think partial data will be return for short peek. Actually all the listener filters are doing similar, when they didn't get enough data, then loop or waiting another event https://github.com/envoyproxy/envoy/blob/e845ee8faf95f21a0c4aad6af78a679ab5f79ff8/source/extensions/filters/listener/proxy_protocol/proxy_protocol.cc#L421-L423

Sorry, I'm not sure I'm following your last question. Does the `tests to common socket handle file` mean we test all the platform(wins, linux, macos) to have same behavior on peek? Probably add to https://github.com/envoyproxy/envoy/blob/main/test/common/network/io_socket_handle_impl_test.cc ?

It sounds like we will define a behavior across all the platforms. So we will use the linux as the behavior standard? It should be easy for short peek, it is hard for all the IoSocketHandle method I guess.

Reply:
Emm...actually, this PR is going to resolve the issue of Windows doesn't support edge triggered event, so we drain all of the data from the socket to emulate the edge triggered_event.

So the goal isn't fully exact same behavior emulation with linux, I don't have confidence it will be exact behavior with linux(but at least, it shouldn't have any behavor can break the linux filter, since all the tests passed).